pysph/sph/wc/gtvf.py

```diff
@@ -0,0 +1,655 @@
+        uhatij[2] = d_what[d_idx] - s_what[s_idx]
+
+        for i in range(d):
+            for j in range(d):
+                d_gradvhat[d_idx*d*d + d*i + j] += Vj * uhatij[i] * DWIJ[j]
+
+
+class DeviatoricStressRate(Equation):
+    r"""**Stress rate for solids**
+
+    From [ZhangHuAdams2017], equation (5),
+
+    .. math::
+            \frac{d \boldsymbol{\sigma}'}{dt} = 2 G (\boldsymbol{\epsilon}
+            - \frac{1}{3} \text{Tr}(\boldsymbol{\epsilon})\textbf{I}) +
+            \boldsymbol{\sigma}' \cdot \boldsymbol{\Omega}^{T} +
+            \boldsymbol{\Omega} \cdot \boldsymbol{\sigma}'
+    where,
+
+    .. math::
+           \boldsymbol{\Omega_{i/j}} = \frac{1}{2}
+           \left(\nabla \otimes \boldsymbol{v}_{i/j} -
+           (\nabla \otimes \boldsymbol{v}_{i/j})^{T}\right)
+
+    .. math::
+           \boldsymbol{\epsilon_{i/j}} = \frac{1}{2}
+           \left(\nabla \otimes \boldsymbol{v}_{i/j} +
+           (\nabla \otimes \boldsymbol{v}_{i/j})^{T}\right)
+
+    see the class VelocityGradient for :math:`\nabla \otimes \boldsymbol{v}_i`
+   """
+    def __init__(self, dest, sources, dim, G):
+        r"""
+        Parameters
+        ----------
+        dim : int
+            Dimensionality of the problem.
+        G : float
+            value of shear modulus
+        """
+        self.G = G
+        self.dim = dim
+        super(DeviatoricStressRate, self).__init__(dest, sources)
+
+    def initialize(self, d_idx, d_asigma):
+        i = declare('int')
+        for i in range(9):
+            d_asigma[d_idx*9 + i] = 0.0
+            printf("%d, %f\n", d_idx*9 + i, d_asigma[d_idx*9 + 1])
+            printf("HI")
+
+    def _get_helpers_(self):
+        return [mat_vec_mult, mat_mult]
+
+    def loop(self, d_idx, s_idx, d_sigma, d_u, d_v, d_w, s_uhat, s_vhat,
+             s_what, d_uhat, d_vhat, d_what, d_asigma, s_sigma, d_gradvhat,
+             s_gradvhat, DWIJ):
+        i, j, d = declare('int', 3)
+        eps, omega, omegaT, sigmai, dvi, dvj = declare('matrix(9)', 6)
+
+        d = self.dim
+        G = self.G
+
+        for i in range(9):
+            sigmai[i] = d_sigma[d_idx*9 + i]
+            dvi[i] = d_gradvhat[d_idx*9 + i]
+            dvj[i] = s_gradvhat[s_idx*9 + i]
+
+        eps_trace = 0.0
+        for i in range(d):
+            for j in range(d):
+                eps[d*i + j] = 0.5*(dvi[d*i + j] + dvj[d*j + i])
+                omega[d*i + j] = 0.5*(dvi[d*i + j] - dvj[d*j + i])
+                if i == j:
+                    eps_trace += eps[d*i + j]
+
+        for i in range(d):
+            for j in range(d):
+                omegaT[d*j + i] = omega[d*i + j]
+
+        tmp1, tmp2 = declare('matrix(9)', 2)
+        mat_mult(sigmai, omegaT, d, tmp1)
+        mat_mult(omega, sigmai, d, tmp2)
+
+        for i in range(d):
+            for j in range(d):
+                d_asigma[d_idx*d*d + d*i + j] = 2*G * eps[d*i + j]
+                d_asigma[d_idx*d*d + d*i + j] += -2*G * eps_trace/3.0
```

Comment:
The trace term should be subtracted only on the diagonals and not for every i and j.

Reply:
thanks, fixed it!